progressive_rollout_demo/lib/main.dart

```diff
@@ -0,0 +1,150 @@
+
+  await Firebase.initializeApp();
+  await assignGroupNumberIfNeeded();
+
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
+      ),
+      home: const MyHomePage(),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key});
+
+  @override
+  State<MyHomePage> createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final _updater = ShorebirdUpdater();
+
+  int? _groupNumber;
+
+  /// Returns the current release version (e.g., 1.0.0+1).
+  Future<String> _releaseVersion() async {
+    final packageInfo = await PackageInfo.fromPlatform();
+    return '${packageInfo.version}+${packageInfo.buildNumber}';
+  }
+
+  /// Fetches the rollout percentage for the current release version
+  /// (e.g., 1.0.0+1) from Firestore.
+  ///
+  /// You can view the Firestore data in the console at:
+  /// https://console.firebase.google.com/u/0/project/progressive-rollout-demo/firestore/databases/-default-/data/~2Frollouts~2Frollouts
+  Future<int?> _fetchRolloutPercentage() async {
+    final collection = await FirebaseFirestore.instance
+        .collection(firestoreCollectionName)
+        .get();
+    final releaseVersion = await _releaseVersion();
+    return collection.docs.first.data()[releaseVersion] as int?;
+  }
+
+  /// Gets the group number (1-100) from shared preferences. We assume elsewhere
+  /// that this has been set before this widget is used.
+  Future<int?> _fetchGroupNumber() async {
+    final prefs = await SharedPreferences.getInstance();
+    return prefs.getInt(groupKey);
+  }
+
+  /// Determines the update track (beta or stable) based on the group number
+  /// and rollout percentage. If a user's group number is less than or equal to
+  /// the rollout percentage, they are assigned to the beta track. Otherwise,
+  /// they are assigned to the stable track. If there is no rollout percentage
+  /// for the current release version, all users are on the stable track.
+  Future<UpdateTrack> _updateTrack() async {
+    final (groupNumber, rolloutPercentage) = await (
+      _fetchGroupNumber(),
+      _fetchRolloutPercentage(),
+    ).wait;
+
+    // If there is no rollout percentage, all users are on the stable track.
+    if (rolloutPercentage == null) return UpdateTrack.stable;
+
+    // If the user's group number is less than or equal to the rollout
+    // percentage, they are on the beta track. For example:
+    // - if the rollout percentage is 25%, users with group numbers 1-25 are on
+    //   the beta track, and the other 75% are on the stable track.
+    // - if the rollout percentage is 100%, all users are on the beta track.
+    return groupNumber! <= rolloutPercentage
```

Comment:
Is it safe to force unwrap groupNumber here?

Reply:
Yep, we assume it's been set in `main` before the app is launched.

Reply:
I don't think that's right. We don't call runApp until after `assignGroupNumberIfNeeded` has completed.

Reply:
`groupNumber` is initialized by `initState` asynchronously no?

Reply:
Oh I was confused by `_groupNumber` and `groupNumber` nevermind 🤦